### ISSUE TITLE
Allow to run docker images as arbitrary user

### ISSUE THREAD

**Issue by hhorak** [sclorg/rhscl2dockerfile#18]
This is a controversial requirement, that already got into some of the OpenShift docker images, but I've still missed the use cases why this is necessary in the docker images..

First valid point is that no docker image should run as `root`, which seems to be something most people agree on IMHO. This may be achieved in daemon images like databases by using user `mysql` or `postgres` that is already created during RPM install in the Dockerfile (`USER mysql`).

In docker images that don't have such a user (ruby, python, nodejs) we probably need to create such a user explicitly in the Dockerfile. The current practice is to call it `default` and use `UID` 1000 or 1001 (where 1000 has better chance to hit the `UID` of user on the host, but otherwise both cases are almost the same).

`docker run -u $(id -u) ...` allows to run the image as arbitrary user already, but what is missing in the image is mapping the `EUID` in the image to the entry in the image's `/etc/passwd` -- is this the issue OpenShift wants to address by features like this https://github.com/openshift/postgresql/commit/1916d06fab2f4e04c7511beaa17c7d6480a0d4d3 ?

AFAICT, this should work nicely in docker in the future, so the above implementation is a temporary solution, right?

There were also ideas that the feature to run the image as arbitrary user has some security benefit, but given we still need to have some non-root user set in the Dockerfile everytime (because it is the only way to ensure `docker exec CONT /bin/sh` is not run as `root`), I don't see how this helps the security. Can somebody elaborate on this?

There is also one important block with the implementation mentioned above -- nss_wrapper is not part of the RHEL so far and we cannot install random packages into RHSCL docker images..

@bparees @mfojtik @mnagy

**Comment by mfojtik** [comment]
> AFAICT, this should work nicely in docker in the future, so the above implementation is a temporary solution, right?

Correct, this should be resolved in Docker by implementing the User Namespaces properly. However, we can still profit from having random UID in Docker image. Especially in cluster, that will grant you more security as you have think about the case where on one node there will be N containers all started by different OpenShift/Kubernetes users.

>  I don't see how this helps the security. Can somebody elaborate on this?

When you do `osc exec container_id` (which use same mechanism as docker exec) as an arbitrary user, you wont get root. But when I think about this more, then I realized that nothing prevent users to do Docker build strategy in OpenShift and have the USER root in Docker file. So this will be more like a 'good practice' than security fix. We want to encourage users to write Dockerfiles that does not require root user in general.

 > There is also one important block with the implementation mentioned above -- nss_wrapper is not part of the RHEL so far and we cannot install random packages into RHSCL docker images..

@cewong can you elaborate here? Is there a way to achieve random user id by not using nss_wrapper?

**Comment by praiskup** [comment]
> @cewong can you elaborate here? Is there a way to achieve random user id by
> not using nss_wrapper?

@mfojtik should not we ping @csrwng for this issue?

**Comment by mfojtik** [comment]
@praiskup corrected, thanks :-)

**Comment by hhorak** [comment]
@mfojtik wrote:
> Correct, this should be resolved in Docker by implementing the User Namespaces properly. However, we can still profit from having random UID in Docker image. Especially in cluster, that will grant you more security as you have think about the case where on one node there will be N containers all started by different OpenShift/Kubernetes users.

I'm still struggling to understand some consequences.. Thinking about a particular example -- e.g. an sti-python image, where we define `USER default` in the dockerfile. A user then works with the docker image as the `default` user (on his workstation), which means the files that will go inside the image will get 1000:1000 as UID:GID like this:
```
-rw-r--r--.  1 1000 1000 ... a_writeable_file
-r--------.  1 1000 1000 ... a_readable_file
```
Then, when it will be run as `docker run -u 12345` in the cluster -- what happens? the docker container won't be able to write to that file..

Maybe I miss the point how the application code gets into the container -- the way I imagine it is that a new layered image is built (in the openshift or on user's workstation), is that right? That way it is possible to do something like:
```
USER root
RUN useradd -u 12345 cluster_user
USER cluster_user
ADD app_code /somewhere
```
Another thought that came to my mind is: in case I want to separate directories that are owned by the same user -- I can have directories names that are not predictable, like:
```
drwxrwxr-x. 12 root root  4096 May 18 17:11 .
drwxrwxr-x. 68 root root  4096 May 15 17:07 ..
drwxr-xr-x.  1 default default  4096 Apr 24 17:59 787b6bc3d7a7e9456140fbd41f11c860d3107285
drwxr-xr-x.  4 default default  4096 May 11 18:50 b58296294b742e94c8c32dbdd677e2373da4f589
```
which seems much cleaner to me than `nss_wrapper` and should do the same thing on the host (do not allow to read content of some directory even if I have unix permission to do it)...

**Comment by csrwng** [comment]
> But when I think about this more, then I realized that nothing prevent users to do Docker build strategy in OpenShift and have the USER root in Docker file. So this will be more like a 'good practice' than security fix. We want to encourage users to write Dockerfiles that does not require root user in general.

@mfojtik - With the use of security contexts in kubernetes, we will always specify the user for a   container. We will never use the one set by the image author. 

The other big reason we want randomized uids is to use external storage. When a user allocates external storage, they will be assigned a globally unique id and permissions will be pre-set with that uid. Images like mysql, postgres, etc should be able to use this storage with the permissions for that user.



> Then, when it will be run as docker run -u 12345 in the cluster -- what happens? the docker container won't be able to write to that file..

@hhorak The approach so far is that in order to make something writeable inside an image we make it world-writable. In a normal system this would be horribly insecure. However, for  a container I would say that this is acceptable given that it's contained.



> @csrwng can you elaborate here? Is there a way to achieve random user id by not using nss_wrapper?

So the reason we're using nss_wrapper is that applications like postgresql insist on calling getpwuid with the current uid. (See my comment here: https://github.com/openshift/postgresql/pull/10#discussion_r29327412) If a valid entry is not returned, initialization fails. For this, we need to have an entry in the passwd file. Beside nss_wrapper one other possible solution would be to generate a passwd file that has all possible uids, but that would be a very large file (~ 2^32 entries ). We could also introduce our own code to override the getpwuid call, but not sure how much better that would be than nss_wrapper.

**Comment by hhorak** [comment]
@csrwng wrote:
> The approach so far is that in order to make something writeable inside an image we make it world-writable. In a normal system this would be horribly insecure. However, for a container I would say that this is acceptable given that it's contained.

@rhatdan Dan, I remember you talking about containers that they actually don't contain -- what do you think about the above -- making some files world-writable in the container?

As I understand it, we are speaking only about one directory to be world-writable, right? In case of mysql that would be `/var/lib/mysql/data` that is mounted from host. But the directory we mount on the host will be also world-writable, which is a security hole rather than improvement.

**Comment by csrwng** [comment]
> As I understand it, we are speaking only about one directory to be world-writable, right? In case of mysql that would be /var/lib/mysql/data that is mounted from host. But the directory we mount on the host will be also world-writable, which is a security hole rather than improvement.

The idea of allowing random uids is that if you mount an external volume into your container you'd secure it for the random user that you're going to run the container as. Nothing that you mount externally would be expected to be world-writeable

**Comment by hhorak** [comment]
@csrwng wrote:
> The idea of allowing random uids is that if you mount an external volume into your container you'd secure it for the random user that you're going to run the container as. Nothing that you mount externally would be expected to be world-writable

Ah, I understand better now probably -- we need the world-writable directory for cases the directory is created during `docker build` and we need to write to the directory during `docker run`. The volume-mounted directories for data will already be owned by proper user and won't be world-writable.

So, in case of mysql image, we only need a directory for config file and for the language-stack images (ruby, python, nodejs) we won't probably need any such world-writable directories. Is that correct?

**Comment by csrwng** [comment]
> So, in case of mysql image, we only need a directory for config file and for the language-stack images (ruby, python, nodejs) we won't probably need any such world-writable directories. Is that correct?

For the most part, yes. In some cases, in the language images you'll need to make a directory writeable, whether it is for config or logs.

**Comment by rhatdan** [comment]
I hate the idea of having random world writable directories inside of containers or outside.  Seems like we might be able to do something with groups.  IE Change the directory to be writable by any UID within the mariadb/mysql group.  Then you could modify the container to add the USER to the group on entry, or better use some kind of external source to provide identity.

**Comment by hhorak** [comment]
One more point for the `nss_wrapper` implementation -- whatever way chose for implementation, we'd end up with `passwd` file writable by the user that runs the container, right? then we'd be able to gain root access very easily:
```
bash-4.2$ echo 'root:some-known-hash:0:0:root:/root'>/var/lib/pgsql/passwd
bash-4.2$ su -     
Password: 
-sh-4.2# id 
uid=0(root) gid=0(root) groups=0(root)
```
This seems like a security flaw to me.

**Comment by csrwng** [comment]
The nss_wrapper only provides a valid entry for getpwuid calls, nothing more. It's a user-level library and doesn't give you any rights other than what you already have. It would need elevated rights to do that. You're still the same user (only getting a different pwd entry when making the getpwuid call). Attempting to do su - with a fake root password will not work.

**Comment by csrwng** [comment]
> I hate the idea of having random world writable directories inside of containers or outside. Seems like we might be able to do something with groups.

@rhatdan For us to give you membership in a group, we'd need to have a pre-configured /etc/passwd file that has all possible uids. (nss_wrapper as I explained above, doesn't do that... you may have it say that you're a member of a group, but that doesn't really make you a member). Would you be in favor of having a preconfigured /etc/passwd with a range of users? or is maybe some other way we can tell the container that the particular user you're running as is a member of a particular group?

**Comment by rhatdan** [comment]
I would like to see the fact that you entered the container, or started the container change the group(s) that you are a member of.  I think this would be the safest case, but I am just brain storming here.

Say I did
 docker run -u 12345 

Then the docker daemon would setup the 12345 UID to be a member of the DockerUserGroup (1000)

Then if I had a directory that could be written by the DockreUserGroup this process could write to it.

Not sure how you would state this in docker terms, but that seems to be the safest way to handle a problem with users only are allowed to manipulate the directories while in the container.

**Comment by hhorak** [comment]
We already have an apparatus to run some commands every-time we enter or run the container (because of enabling SCL environment) -- it's called `container-entry` script. So, if I understand @rhatdan 's way correctly, this feature could be implemented by running something like `usermod -a -G default $(id -u)` during in the `container-entry` + directories we want to write to would have `0775` as permissions and `root:default` as owner..

But that won't solve the issue with postgresql daemon calling `getpwuid`, so we would also have to run `useradd -u 1234 containeruser`.

Another question is how to call those commands as non-root user? Running some script that would handle it safely using `sudo` and providing some `sudoers` magic doesn't seem very nice to me at all.

**Comment by csrwng** [comment]
Yup, exposing sudo even if narrow seems dangerous. Ideally this would be done by the program that runs the container (for example, bind mount an external /etc/passwd with the user properly configured)

**Comment by mfojtik** [comment]
@csrwng who will provide the external /etc/passwd? Secret?

**Comment by csrwng** [comment]
in our case it would have to be done by kubernetes... maybe can be accomplished by a custom volume type (just brainstorming)

**Comment by rhatdan** [comment]
I don't want to modify the contents of the container, but to treat it as read/only.  But a privileged process would be required to change the group identity of the process entering the container.  

Identify could be provided via an external source with proper setup of nsswitch.  So that the calls to getpw* would return the correct information

**Comment by hhorak** [comment]
@rhatdan wrote:
>I don't want to modify the contents of the container, but to treat it as read/only. But a privileged process would be required to change the group identity of the process entering the container.
Identify could be provided via an external source with proper setup of nsswitch. So that the calls to getpw* would return the correct information

Sorry, I don't understand completely, can you be more specific, what it would mean for the image content itself?

**Comment by rhatdan** [comment]
Running usermod inside the container to manipulate the /etc/passwd and /etc/group file.  If the identity was being provided by an external source like sssd socket leaked into the container, or another object, then you could get the getpw calls to work correctly and you could call the appropriate calls to change the grouplist of the process entering the container.

**Comment by hhorak** [comment]
@rhatdan wrote:
> Running usermod inside the container to manipulate the /etc/passwd and /etc/group file.

...by privileged process... which was said to be no-go for openshift.

@rhatdan wrote:
> If the identity was being provided by an external source like sssd socket leaked into the container, or another object, then you could get the getpw calls to work correctly and you could call the appropriate calls to change the grouplist of the process entering the container.

For general purpose image we can't depend on it, or is this something which would work for openshift, @csrwng ? Otherwise I still don't see a feasible solution..

**Comment by rhatdan** [comment]
I am just throwing out ideas here.  I would suggest we think about having docker put a user into a standard group when entering a container.  Not a privileged process.  Maybe something you could specify on the command line.

docker run/exec -u 1234 -g 23567 

Then docker would stick the user process into the 23567 group while it is running in the container.

Now you could make the content that you want to give access to the user to group 23567 with rwx.

**Comment by csrwng** [comment]
@hhorak yup I don't think we can rely on an external source for identity
@rhatdan I like the idea. User namespace support would also solve the problem for us.

**Comment by rhatdan** [comment]
Has anyone tried 

docker run/exec -u 1234:23567  

It looks from the man page like this should work.

I will still looking at adding --grouplist to equivalent of /etc/group.
groups
dwalsh wheel dialout docker bumblebee rvm

**Comment by hhorak** [comment]
@rhatdan ha, `docker run -u 1234:23567` works! I got 23567 as GID in the container and after `docker exec` (which doesn't allow to set `-u`) I got the same UID/GID..

@csrwng I think we don't still know how to solve missing entry in `/etc/passwd` for postgresql image, which needs to have some name found for the UID, right?

**Comment by rhatdan** [comment]
Looks like libcontainer already has concept of AdditionalGroups.

```
$ grep -r AdditionalGroups  .
./init_linux.go:	suppGroups := append(execUser.Sgids, config.Config.AdditionalGroups...)
Binary file ./nsinit/nsinit matches
./configs/config.go:	// AdditionalGroups specifies the gids that should be added to supplementary groups
./configs/config.go:	AdditionalGroups []int `json:"additional_groups"`

$ grep -r suppGroups  .
./init_linux.go:	suppGroups := append(execUser.Sgids, config.Config.AdditionalGroups...)
./init_linux.go:	if err := syscall.Setgroups(suppGroups); err != nil {
Binary file ./nsinit/nsinit matches
```

**Comment by csrwng** [comment]
@hhorak - yes, we still need some way to get a valid pwd entry. If we want to avoid nss_wrapper we could write our own nss provider, not sure if that's a better solution from your point of view.

**Comment by hhorak** [comment]
@csrwng I don't see any good enough solution for that so far. Less hackish way mentioned so far would be the pre-generated `passwd` file. Since it is a temporary solution, can we limit the `UID` for some reasonable range, like 1000-2000?

**Comment by praiskup** [comment]
The supplemental group idea is promising, but well, when we know that we can
change the GID, where is the problem?  Do we actually need overriding the
getpwid()?)

Tested against https://github.com/fedora-cloud/Fedora-Dockerfiles/pull/105 :

Simply do something like this on host:

```
$ RANDOM_GID=`echo "$RANDOM % 999 + 1001" | bc`
$ mkdir testdata
$ sudo chcon system_u:object_r:svirt_sandbox_file_t:s0 testdata
$ sudo chown root:"$RANDOM_GID" testdata/
$ sudo chmod 070 testdata
$ sudo docker run -ti \
    -e POSTGRESQL_CONTAINER_OPTS="pgdata = /var/lib/pgsql/data/data" \
    -v `pwd`/testdata:/var/lib/pgsql/data \
    -u 26:"$RANDOM_GID" \
    docker-registry.usersys.redhat.com/praiskup/fedora-postgresql:latest
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

creating directory /var/lib/pgsql/data/data ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
creating configuration files ... ok
creating template1 database in /var/lib/pgsql/data/data/base/1 ... ok
initializing pg_authid ... ok
initializing dependencies ... ok
creating system views ... ok
loading system objects' descriptions ... ok
creating collations ... ok
creating conversions ... ok
creating dictionaries ... ok
setting privileges on built-in objects ... ok
creating information schema ... ok
loading PL/pgSQL server-side language ... ok
vacuuming database template1 ... ok
copying template1 to template0 ... ok
copying template1 to postgres ... ok
syncing data to disk ... ok

WARNING: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    postgres -D /var/lib/pgsql/data/data
or
    pg_ctl -D /var/lib/pgsql/data/data -l logfile start

error: /var/lib/pgsql/data/data dir does not seem to be initialized
LOG:  redirecting log output to logging collector process
HINT:  Future log output will appear in directory "pg_log".
^C$
```

There are three things we should take into account; 

- the data directory on host must be owned by some UID other than 26=postgres
  (root seems to be reasonable choice as you refuse users to run root
  processes),

- the directory should have ideally 'chmod ao-rwx' set and

- you need to reconfigure the data directory to be in subdir like
  `/var/lib/pgsql/data/SUBDIR`.  That does not seem to be a problem based on the
  previous example (Kubernetes has all the info actually needed to automatize
  that).

(and please ignore that 'error:' message, the change of pgdata location was
not yet tested, I'll fix asap)

**Comment by csrwng** [comment]
@praiskup we still need a random uid, not just gid. The reason is that the data volume we mount will likely be on external storage with a directory that's already owned by a previously provisioned uid for the app.

**Comment by praiskup** [comment]
@csrwng wrote
> @praiskup we still need a random uid, not just gid. The reason is that the
> data volume we mount will likely be on external storage with a directory
> that's already owned by a previously provisioned uid for the app.

Sorry, can you give me an example?

**Comment by csrwng** [comment]
@praiskup - Sorry I was thinking of a storage directory owned by the random user, with typical permissions like 0740. But now I think about it, I don't see why we couldn't set it up with the same permissions (0070) you have in your example.

**Comment by hhorak** [comment]
@praiskup thanks, that's great approach. the only piece I don't understand is why we need to use a subdirectory -- `/var/lib/pgsql/data/data` and not just `/var/lib/pgsql/data`?

**Comment by praiskup** [comment]
Because while mounting the /var/lib/pgsql/data into the container (with that permissions),
even the container itself is not able to touch the directory's metadata.  That means,
postgresql fails on "fixing permission" error.

**Comment by hhorak** [comment]
@csrwng Since this may be solved outside of images very cleanly and we're still speaking about a temporary solution until docker has proper User Namespaces, can we close this issue and you guys will hack it into openshift once for every image?

**Comment by csrwng** [comment]
@hhorak @praiskup The one thing that this still doesn't solve for us is that in a multi-tenant environment we get the security benefit of running processes(containers) as different users when using random uids. Using this approach would mean that multiple containers (owned by different tenants) run with the same id.

**Comment by praiskup** [comment]
@csrwng you talk about some situation, where (i would rather say arbitrary
than random) GID does not help?  Can you describe the multi-tenant
scenario please?

**Comment by hhorak** [comment]
@csrwng I see it might not be perfect for openshift's use cases, but I actually didn't get whether it is a good enough compromise or not. We're still looking for a solution which will work for RHSCL images (where we simply can't install epel packages).

**Comment by csrwng** [comment]
@praiskup - The main scenario is you compromise the container running your own pgsql... find a vulnerability and escape the container. As the pgsql user you can see and potentially compromise other processes you're running which is everybody else's container running pgsql. 

@hhorak - let me discuss with the team and evaluate the solutions. So far the options seem to be 1) run with a pre-generated /etc/passwd that can work for a range of users or 2) use the random gid solution that @praiskup suggested. Right?

**Comment by praiskup** [comment]
@csrwng wrote
> As the pgsql user you can see and potentially compromise other processes
> you're running which is everybody else's container running pgsql.

Could SELinux (MCS) help here?  Seems like that is something virtualization
uses to solve the same issue.

**Comment by hhorak** [comment]
@csrwng wrote:
> @hhorak - let me discuss with the team and evaluate the solutions. So far the options seem to be 1) run with a pre-generated /etc/passwd that can work for a range of users or 2) use the random gid solution that @praiskup suggested. Right?

Basically yes, plus the SELinux idea above (@rhatdan, does it sound like something what can help here?).

Anyway, please mind that we need to know the resolution quite soon in order to be able to incorporate the changes into RHSCL images.

**Comment by hanzz** [comment]
@csrwng wrote
> @hhorak The approach so far is that in order to make something writeable inside an image we make it world-writable. In a normal system this would be horribly insecure. However, for a container I would say that this is acceptable given that it's contained.

I already wrote it in [my comment in #17](https://github.com/sclorg/rhscl2dockerfile/issues/17#issuecomment-105469605), but the way how that works for sti-php means that someone hacking httpd would be able to change the configuration of the httpd (or would be able to change the logs). I think this is not good solution for httpd at least. Can you prove that I'm wrong here?

Unfortunately I don't have better solution right now.

**Comment by hanzz** [comment]
I got an idea how to fix this issue. Note that I did not try this in practice, this is pure theory now.

Can we have entry-point executable binary owned by root:root with setuid bit set? Openshift would start the container with non-root user, that setuid bit binary would do the "chown" of particular directories (/etc/httpd or /var/log/httpd in httpd case), do the preparation needed to be done by the Openshift and then switch to non-root and execute the service?

That way, user running "docker exec" or "osc exec" would get the non-root shell. You would not have to do the bad chmod to allow world-writable /etc/httpd and stuff. You would just do the right "chown" of the directories which really should be writable by the service.

We would have to be sure that there are no bugs in that setuid binary. It could probably just be a wrapper to call bash script owned by root:root with strict permissions.

If you would like to have non-root with httpd, it could also be possible to create two non-root users, one which would be able to edit /var/log/httpd and other httpd directories and one which would be used to run the httpd workers (which should not be able to edit /var/log/httpd).

**Comment by hhorak** [comment]
Well, suid binary could also be a solution for creating passwd entry for the arbitrary user -- it would be run during image initialization and would delete itself in the end to prevent any other possible abuse.. But it is still too hacky and we would have to be very very careful with the suid. And we don't have gcc inside the container, so how to get the binary there?

**Comment by praiskup** [comment]
Hmm, suid binary is something which comes to mind when some hack needs to
be done .. that binary would require to be pre-packaged as a part of httpd
package, that is probably doable - special audit would need to be done,
etc.   Then you usually want to migrate to capabilities (already discussed
somewhere, can't remember where).  But, IMHO, those are things that should
be solved by environment than by ugly hacks per-container :/

**Comment by hanzz** [comment]
@hhorak wrote
> Well, suid binary could also be a solution for creating passwd entry for the arbitrary user -- it would be run during image initialization and would delete itself in the end to prevent any other possible abuse.. But it is still too hacky and we would have to be very very careful with the suid.

It's hacky, but still more cleaner than "chmod -R og+rwx /opt/rh/php55/root/etc" in DockerFile. Personally I don't image better way which fulfils Openshift requests and security problems I have pointed out earlier in this discussion.

> And we don't have gcc inside the container, so how to get the binary there?

Could we add it as some new "scl-docker-run" RPM into RHSCL-2.0 (Yeah, that does sound insane... :), but please consider it)? This is probably the only way how to test that stuff properly, have it in single place and be able to update it easily.

@praiskup wrote
> But, IMHO, those are things that should be solved by environment than by ugly hacks per-container :/

I fully agree with you here. The best way is to wait for user namespace support, but as we have discussed earlier, this is no-go for Openshift and they won't use RHSCL if we do not support their use-cases. Currently, their way of implementing the non-root user looks much more suspicious and dangerous to me than the proposed suid binary for initialization.

Note that I still haven't tried that with the real container, so maybe this way cannot work as I expect, but so far I don't see any reason why this should not work.

**Comment by hhorak** [comment]
@hanzz We would have to talk about it with security guys, but first we should at least know if it really works. Honzo, can you look at it in practice then and try out if it work?

**Comment by hanzz** [comment]
Done that already this morning and it seems it's working for me. Lets discuss it in person and write some follow-up here later.

**Comment by hanzz** [comment]
OK, here's the proof of concept: https://github.com/hanzz/scl-container-init

We could use capabilities later if there will be some consensus on what root features should be available to that "currently suid binary". I presume that Openshift user cannot redefine entry-point of the image, otherwise we would have to clear the environment in the scl-container-init binary.

**Comment by hanzz** [comment]
> If you would like to have non-root with httpd, it could also be possible to create two non-root users, one which would be able to edit /var/log/httpd and other httpd directories and one which would be used to run the httpd workers (which should not be able to edit /var/log/httpd).

This is not true, we would really need root and non-root user anyway unless someone finds a way how non-root user could drop the priveleges to another non-root user :).

**Comment by hhorak** [comment]
I like this approach since it is general (we can use it in all images, not only in httpd) and as soon as we have User Namespaces in docker we may get rid of it quite nicely.

**Comment by hhorak** [comment]
@ppisar @remi @sYnfo @rkuska If the perl/python/php modules are installed by calling such a suid binary, they will be installed as root and won't be possible to be rewritten by user running the application. Does it seem sane for you?

**Comment by hhorak** [comment]
@humaton @strzibny FYI ^

**Comment by notroj** [comment]
@mfojtik @rhatdan @csrwng Can you guys confirm this approach is actually viable?  It requires we can can setuid to root (once) within the container.  You don't intend to lock that down?

**Comment by ppisar** [comment]
On Thu, May 28, 2015 at 04:10:05AM -0700, Honza Horak wrote:
> @ppisar @remi @sYnfo @rkuska If the perl/python/php modules are installed by
> calling such a suid binary, they will be installed as root and won't be
> possible to be rewritten by user running the application. Does it seem sane
> for you?
>
Installing an application under one user and running it under another user is
good. This ensures the running application will not be able to modify its
code or configuration.

This solves the final phase of running an application. But I worry it creates
annother problem in previous phase of installing the application.

If I understand the STI correctly, the application installation means running
a container. And with this SUID entrypoint you would get a docker container
with fixed UID which is exactly this issue (Allow to run docker images as
arbitrary user) is against.

Also don't forget that installing Perl application means downloading Perl
modules from the Internet and executing their build scripts. Effectively, you
would get root access `anybody' to your system at time of installing the
application. That make me nervous. However, that's nothing new to everybody
doining this already without container technology, so we can ignore this
aspect.

**Comment by hanzz** [comment]
@ppisar wrote
> This solves the final phase of running an application. But I worry it creates
annother problem in previous phase of installing the application.
>
> If I understand the STI correctly, the application installation means running
a container. And with this SUID entrypoint you would get a docker container
with fixed UID which is exactly this issue (Allow to run docker images as
arbitrary user) is against.

You could still download/build the Perl modules using the random UID, but using the suid binary during the container initialization, you should be able to "chown root:root -R my_perl_modules" when the installation phase is completed.

I agree that the usage of shared root user even for suid binary is not good and the best solution would be the user namespace, but I find it less problematic than allowing applications to edit their own source code or configuration when something bad happens to them. The script running as root would be known and can be verified and it would be running only during the initialization for limited time. I'm not sure we can do that differently when the only user being able to change the container from inside is shared root.
